docs: add `SubresourceIntegrityPlugin` (#9302)
web-infra-dev/rspack: website/docs/en/plugins/rspack/subresource-integrity-plugin.mdx

import { ApiMeta } from '@components/ApiMeta.tsx';

# SubresourceIntegrityPlugin

<ApiMeta specific={['Rspack']} addedVersion="1.2.4" />

The `rspack.experiments.SubresourceIntegrityPlugin` is a plugin for enabling Subresource Integrity in Rspack.

> > [Subresource Integrity](http://www.w3.org/TR/SRI/) (SRI) is a security

Reviewer: How about copying the "What is SRI" section from the Rsbuild documentation:

https://github.com/web-infra-dev/rsbuild/blob/71ea28b23f37c25303017870740bbb495b805683/website/docs/en/config/security/sri.mdx?plain=1#L16

> feature that enables browsers to verify that files they fetch (for
> example, from a CDN) are delivered without unexpected
> manipulation.

## Features

### Support for HTML Plugin

> The plugin supports integration with [`HtmlRspackPlugin`](/plugins/rspack/html-rspack-plugin.mdx) and [`html-webpack-plugin`](https://github.com/jantimon/html-webpack-plugin). It will automatically set the `integrity` and `crossorigin` attributes for the injected tags.

Reviewer: ```suggestion
The plugin supports integration with [`HtmlRspackPlugin`](/plugins/rspack/html-rspack-plugin) and [`html-webpack-plugin`](https://github.com/jantimon/html-webpack-plugin). It will automatically set the `integrity` and `crossorigin` attributes for the injected tags.
```

### Support for Code Splitting

> The plugin supports code splitting. When you use dynamic imports, the plugin will automatically set the `integrity` and `crossorigin` attributes while generating the chunk loading tags.

Reviewer: ```suggestion
The plugin supports code splitting. When you use dynamic imports, the plugin will automatically set the `integrity` and `crossorigin` attributes for the generated chunk loading tags.
```

## Usage

You can use the plugin by importing it as an experimental plugin from `@rspack/core`:

```js title="rspack.config.mjs"

> import { experiments: { SubresourceIntegrityPlugin } } from "@rspack/core";

Reviewer: Is this a valid syntax?

![image](https://github.com/user-attachments/assets/2f2ad2ba-44e2-467c-b7ba-041cd185af82)

Author: Cursor is a lier!

```

> or

Reviewer: ```suggestion
Or:
```

```js title="rspack.config.cjs"
const {
  experiments: { SubresourceIntegrityPlugin },
} = require('@rspack/core');
```

### Recommended Rspack Configuration

> The `crossOriginLoading` option is required for SRI to work:

Reviewer: ```suggestion
The [output.crossOriginLoading](/config/output#outputcrossoriginloading) option is required for SRI to work:
```

    href="<%= htmlWebpackPlugin.files.css[index] %>"
    integrity="<%= htmlWebpackPlugin.files.cssIntegrity[index] %>"
    crossorigin="<%= webpackConfig.output.crossOriginLoading %>"
  />
<% } %>
```

### Without HTML Plugin

The `integrity` can also be obtained from `stats.assets`. For example:

```js
compiler.plugin('done', stats => {
  const integrityValues = stats
    .toJson()
    .assets.map(asset => [asset.name, asset.integrity]);
});
```

:::tip
Note that when you add the `integrity` attribute on your `link` and
`script` tags, you're also required to set the `crossorigin`
attribute. It is recommended to set this attribute to the same value
as the Rspack `output.crossOriginLoading` configuration option.
:::

## Options

#### hashFuncNames

- **Type:** `Array<"sha256" | "sha384" | "sha512">`
- **Default:** `["sha384"]`

An array of strings, each specifying the name of a hash function to be
used for calculating integrity hash values. Only supports `sha256`, `sha384`, and `sha512` yet.

> See [SRI: Cryptographic hash functions](http://www.w3.org/TR/SRI/#cryptographic-hash-functions) for more details.

#### enabled

- **Type:** `"auto" | boolean`
- **Default:** `"auto"`

> `auto` is the default and means to enable the plugin when the [Rspack
> mode](/config/mode.mdx) is `production` or `none` and disable it when it is `development`. `true` means to enable the plugin and `false` means to disable the plugin.

Reviewer: ```suggestion
- `auto` is the default value, which means the plugin is enabled when [Rspack mode](/config/mode) is `production` or `none`, and disabled when it is `development`.
- `true` means the plugin is enabled in any mode.
- `false` means the plugin is disabled in any mode.
```
